pkg/config/config.go

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"fmt"
+	ioutil "io/ioutil"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/config/v1alpha1"
+)
+
+// ControllerConfiguration defines the functions necessary to parse a config file
+// and to configure the Options struct for the ctrl.Manager
+type ControllerConfiguration interface {
+	runtime.Object
+
+	// GetControllerConfiguration returns the versioned configuration
+	GetControllerConfiguration() (v1alpha1.ControllerConfigurationSpec, error)
+}
+
+// DeferredFileLoader is used to setup the file loader for manager.Options
+// and a ComponentConfig type
+type DeferredFileLoader interface {
+	GetControllerConfiguration() (v1alpha1.ControllerConfigurationSpec, error)
+	AtPath(string) DeferredFileLoader
+	OfKind(ControllerConfiguration) DeferredFileLoader
+	WithScheme(*runtime.Scheme) DeferredFileLoader
+}
+
+// FromFile will set up the deferred file loader for the configuration
+// this will also configure the defaults for the loader if nothing is
+//
+// Defaults:
+//   Path: "./config.yaml"
+//   Kind: GenericControllerConfiguration
+func FromFile() DeferredFileLoader {
+	scheme := runtime.NewScheme()
+	utilruntime.Must(v1alpha1.AddToScheme(scheme))
+	return &deferredFileLoader{
+		path:   "./config.yaml",
+		kind:   &v1alpha1.ControllerConfiguration{},
+		scheme: scheme,
+	}
+}
+
+// deferredFileLoader is used to configure the decoder for loading controller
+// runtime component config types
+type deferredFileLoader struct {
+	kind   ControllerConfiguration
+	path   string
+	scheme *runtime.Scheme
+	once   sync.Once
+	err    error
+}
+
+// GetControllerConfiguration will use sync.Once to set the scheme
+func (d *deferredFileLoader) GetControllerConfiguration() (v1alpha1.ControllerConfigurationSpec, error) {
+	d.once.Do(d.loadFile)
+	return d.kind.GetControllerConfiguration()
```

Comment:
looks like error is never used here, it's supposed to be, right?

Reply:
Opps, looks like you might be right.

Reply:
Added a check also noticed that I messed up the test that was supposed to catch this 🤦 . Should be fixed now.

Reply:
```
if d.err != nil {
  return v1alpha1.ControllerConfigurationSpec{}, d.err
}
```

and updated https://github.com/kubernetes-sigs/controller-runtime/pull/891/files#diff-29cea1b5b831c8655c7155f43e2367cec73e66d1e338f8b3c7877a2f339b8811R32

Reply:
I'm not sure why this isn't showing resolved and the snippet is showing old code - updated code https://github.com/kubernetes-sigs/controller-runtime/pull/891/files/#diff-a3d824da3c42420cd5cbb0a4a2c0e7b5bfddd819652788a0596d195dc6e31fa5R75-R80